### ISSUE TITLE
Line transfer/scattering options

### ISSUE THREAD

**Issue by kslong** [sirocco-rt/sirocco#365]
Currently there are at least two XXX inline comments having to do with line transfer and scattering modes.  The two are interrelated because when you read in the line transfer mode (line 314 of setup_domains.c) what happens is you assign three variables
```
  rdint
    ("Line_transfer(0=pure.abs,1=pure.scat,2=sing.scat,3=escape.prob,4=anisotryopic,5=thermal_trapping,6=macro_atoms,7=macro_atoms+aniso.scattering)",
     &geo.line_mode);
```

* geo.scatter_mode
* geo.line_mode
* geo.rt_mode

There are currently 9 separate options for line transfer mode in the rdint statement, although in fact what happens almost immediately is that geo.line_mode gets winnowed down to 3 values as the other variables are assigned.

The first XXX comment states:

```
/* ksl XXX  This approach is inherently dangerous and should be fixed.  We read in the line mode but then
 * change the number to accommodate a line mode and the way scattering is treated.  We should define 
 * a new variable which we keep as is, and use it to define geo.line_mode and geo.scatter_mode. */
```

Buried in anisowind, there is another XXX question about all of this, which as I have annotated it says:

```
/** 
 * @bug This is a note related to an XXX comoment which read as follows:
 * Factor of 2 needed because the interval is from -1 to 1
 * XXX It's definitely needed to make a uniform distribution work
 * but is this reason really right
 *
 * 180415 - ksl - My investigation of this indicates that the cdf was generated
 * from -1 to 1, and furthermore that cdf_gen_from_function which is used to
 * create the cdf always sets the norm to 1.  So it is not clear to me why
 * this is the reweghting.  It is also nto clear to me why the reweighting to
 * be less than a factor of 2, which is what if stateement requries.  Basically,
 * I do not understand what is going on here at all.
 */
```
This is for a mode of scattering called SCATTER_MODE_ANISOTROPIC.  My guess is that this was used in the original Long & Knigge paper.  It is run whenever one enters 4 for the transfer mode, but we do not seem to be using it currently.  All of our examples use either 3 or 5, which correspond to either isotropic scattering or what we call SCATTER_MODE_THERMAL. The term SCATTER_MODE_THERMAL (which I think James implemented) is not really THERMAL as far as I can tell (as it does not do anything involving temperature).  Rather is seems to be a different way of implementing anisotropic scattering

That said a quick test of what we is called anisotropic scattering and what is called thermal yielded different results for the two choices:

![test](https://user-images.githubusercontent.com/2530901/38772988-d6c79760-4008-11e8-9805-6c60536386ba.png)

So the questions to resolve are:

*What is the difference between the "old" anistropic mode and the "new" one, if that is a fair description of the problem
* What line transfer modes do we actually want to retain, and why?
* If we want to retain the original anisotropic line transfer mode, can we answer the XXX question

**Comment by jhmatthews** [comment]
Hi Knox. This has been documented in a few places:
* https://github.com/agnwinds/python/wiki/Line-Transfer-and-Scattering
* https://sites.google.com/site/agnwinds/current-work-in-progress/scattering-modes
* https://github.com/agnwinds/python/issues/82

The second one was me trying to figure out what was going on. The first is what eventually ended up in the readthedocs input guide. The third is an improvement that was made to the rejection method to speed up things.

From memory, SCATTER_MODE_THERMAL was implemented by @ssim before my time and thermal refers to "thermal trapping", by which I think we mean that because the line profile actually has a thermal width, the Sobolev surface is not really infinetesimal and a photon can undergo a series of scatters and will escape along a preferred direction (max dv/ds). The rejection method samples the probability distribution of this by generating random directions. The improvement in the 3rd link was a renormalisation of this method to avoid performance issues.

As I describe in the second link above, the SCATTER_MODE_ANISOTROPIC is an approximation.

I don't see any particular reason to keep the approximate treatment. We accidentally used the isotropic mode in Nick's paper, and all my papers have used SCATTER_MODE_THERMAL.

**Comment by jhmatthews** [comment]
What's your opinion here @kslong ? 

A related issue is one of the XXX comments in emission.c, which should be addressed once we've decided whether the "analytic" anisotropic mode should be kept. 

```C
	  /*
	     Determine the direction of the photon
	     ?? Need to allow for anisotropic emission here
	     JM 1406 -- XXX I think there's a mistake here. I believe this should be
	     if (p[n].nres < 0 || p[n].nres > NLINES || geo.scatter_mode == SCATTER_MODE_ISOTROPIC)
	     to allow for isotropic BF continuum emission
	   */
	  nnscat = 1;
	  if (p[np].nres < 0 || geo.scatter_mode != SCATTER_MODE_ANISOTROPIC)
	    {
/*  It was either an electron scatter so the  distribution is isotropic, or it
was a resonant scatter but we want isotropic scattering anyway.  */
	      randvec (p[np].lmn, 1.0);	/* The photon is emitted isotropically */
	    }
	  else if (geo.scatter_mode == SCATTER_MODE_ANISOTROPIC)
	    {			// It was a line photon and we want anisotropic scattering 

/* -1. forces a full reinitialization of the pdf for anisotropic scattering  */

	      randwind (&p[np], p[np].lmn, wmain[icell].lmn);
	    }
	  else if (geo.scatter_mode == SCATTER_MODE_THERMAL)
	    {			// It was a line photon and we want anisotropic scattering 
	      randwind_thermal_trapping (&p[np], &nnscat);
	    }
	  p[np].nnscat = nnscat;
```

The above looks plain wrong for line photon generation anyway - it implies that if we select SCATTER_MODE_THERMAL then the final if statement is never entered even though it should be.

**Comment by kslong** [comment]
I think there are two issues here, one directly associated with this topic, and one associated more generally with “historical” modes in Python.

I believe that SCATTER_MODE_ANISOTROPIC was probably the mode use in Long & Knigge.  But as you say it may not be exact if the radiation field is complex, whereas we think that the MC approach in SCATTER_MODE_THERMAL should be accurate in all situations.  Obviously it would be useful to know how the two compare in some practical situations.  If there is not difference, then we could simply remove it.  If there is, then I guess it ought to be documented somehow.

All in all we need to figure out how to deprecate and remove things.  (Another area where we have this problem is ionization modes).  I think we need to have a roundtable discussion on that soon.

**Comment by jhmatthews** [comment]
The above issue regarding the if statement has been fixed by Christian in pull request #402  
The general issue raised by Knox is address in pull request #404 

Merging of these will close the issue.
